swarm/network/stream/delivery.go

```diff
@@ -216,19 +217,37 @@ type ChunkDeliveryMsgRetrieval ChunkDeliveryMsg
 type ChunkDeliveryMsgSyncing ChunkDeliveryMsg
 
 // chunk delivery msg is response to retrieverequest msg
-func (d *Delivery) handleChunkDeliveryMsg(ctx context.Context, sp *Peer, req *ChunkDeliveryMsg) error {
+func (d *Delivery) handleChunkDeliveryMsg(ctx context.Context, sp *Peer, req interface{}) error {
 	var osp opentracing.Span
 	ctx, osp = spancontext.StartSpan(
 		ctx,
 		"handle.chunk.delivery")
 
 	processReceivedChunksCount.Inc(1)
 
+	var msg *ChunkDeliveryMsg
+	var mode chunk.ModePut
+	switch r := req.(type) {
+	case *ChunkDeliveryMsgRetrieval:
+		msg = (*ChunkDeliveryMsg)(r)
+		// do not sync if peer that is sending us a chunk is closer to the chunk then we are
+		peerPO := chunk.Proximity(sp.ID().Bytes(), msg.Addr)
+		po := chunk.Proximity(d.kad.BaseAddr(), msg.Addr)
+		if peerPO > po {
+			mode = chunk.ModePutRequest
+		} else {
+			mode = chunk.ModePutSync
```

Comment:
This is a delivery request. If we got the chunk from the node that is closer to the chunk, we do not need to put in the storage to sync it. I am not sure what is a `syncing message`? We are deciding here should we put this chunk in syncing index or not.

Reply:
sry i know i said this but it should only be the case for chunks outside our depth.
chunks within the area of responsibility should always sync

Reply:
@zelig thanks for the explanation, I may have understood wrongly. Could you check if current code handles various cases correctly?

Reply:
@zelig I was discussing this with @janos today and I have a few questions in this matter:
1. the chunk comes from somewhere, that somehwere should already be the so called "area of responsibility" (let's assume that we are talking about push syncing right now) - aren't we intervening in the process of syncing that a chunk should arrive correctly to the correct NN? also aren't we generating redundancy which is bigger than the premise of `redundance of degree 𝑛 in the network`? this is to say even more risky since area of responsibility is not symmetric between peers

2. I find it very hard to define a test case for this and so I would assume this is a premature optimisation. I think that right now with all that has to do with routing and syncing we should stick to correctness over performant, moreover, our code should be loyal to the spec we are writing. we shouldn't optimise anything at this point (IMO)

Reply:
@justelad 
1. no its not generating more redundancy the rest i dont understand
2. it is easy to construct test cases. this is about correctness